Find root build tasks when calculating an included build task graph

jjohannes opened the pull request:
Fixes the test cases that are part of this PR.

The actual change is in this commit: 981bda426c1c1b8544722b58d87070132ece1a91

see #6229.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":724646317,"teamCityBuildId":"39313794","teamCityBuildHeadRef":"a090262095e4ec70b1c8114288820e0c4c4eec51","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39313794) for you.

jjohannes commented:
@rieske @big-guy I am not sure if this is good enough for now. The type hierarchy of "BuildState" and the logic connected to it is a mess. I think without a huge refactoring doing the distinction between root and included builds in `DefaultIncludedBuildTaskGraph` is the right thing to do. Maybe some details can be improved (see the other comments).

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":724648676,"teamCityBuildId":"39314855","teamCityBuildHeadRef":"2589f9b7b3e57b936287daeb54ce9c83abf68230","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already cancelled the [old build](https://builds.gradle.org/viewLog.html?buildId=39313794) and triggered a new [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39314855) for you.

big-guy commented:
@jjohannes Yeah... I think we should live with the restriction that the root build is special for now. The workaround is to put both builds into a composite with a new empty root. 

We could refactor the way we handle these things to make the root build another "regular" included build and move the root-build special behavior to the build tree idea we had. I don't think we should make the included build controller start treating the root build special.

jjohannes commented:
@big-guy why did you close this PR? I think the solution goes into the right direction and just needs some more exploration if it can be improved. Although I don't see why we could also use it as is to make this possible at least (even if it can be improved performance wise later). Can you please share if you think there are any "dangers" in going with this solution for now (given that it works with more test cases added)?

If anything, we should at least merge the test cases as "to-be-fixed".

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":728985057,"teamCityBuildId":"39530354","teamCityBuildHeadRef":"fc8958a82769579f373711800246a6a82b1b163a","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39530354) for you.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":729037216,"teamCityBuildId":"39533442","teamCityBuildHeadRef":"7d08a638e8569e753dcbdff79a6a2520c81f0144","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already cancelled the [old build](https://builds.gradle.org/viewLog.html?buildId=39530354) and triggered a new [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39533442) for you.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":729045165,"teamCityBuildId":"39534002","teamCityBuildHeadRef":"f45b93aa651009632a92760b1397c04d12937be9","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39534002) for you.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":729052868,"teamCityBuildId":"39534497","teamCityBuildHeadRef":"3783c555f4fa5382203605b1c0d2f07f1d9ba0cf","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39534497) for you.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":729084350,"teamCityBuildId":"39535854","teamCityBuildHeadRef":"9bbb571374ea92c42428d754f038329662266892","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already cancelled the [old build](https://builds.gradle.org/viewLog.html?buildId=39534497) and triggered a new [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39535854) for you.

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":730551543,"teamCityBuildId":"39610126","teamCityBuildHeadRef":"f09f403d94c114e872a7a176c5ef5d130e4b248d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39610126) for you.

jjohannes commented:
@bot-gradle test this

jjohannes commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":731027251,"teamCityBuildId":"39630123","teamCityBuildHeadRef":"a7fd3e542916f7cdbe352f1e988994b7f494058e","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewLog.html?buildId=39630123) for you.

jjohannes commented:
I addressed the concerns I had myself and now the proper state of root build tasks is determined.
Tasks are only added to the root build graph as additional entries if they are not already part of the graph. Also this is done on a "lower level" now, so that no complete reevaluation of everything is triggered.
I also extended the test coverage.

In this [sample build](https://github.com/jjohannes/idiomatic-gradle) you can now run `../../gradlew build` in `product/ig-api` which was not possible without this fix.

bamboo commented on subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultIncludedBuildTaskGraph.java:
```diff
@@ -44,16 +64,44 @@ public void awaitTaskCompletion(Collection<? super Throwable> taskFailures) {
 
     @Override
     public IncludedBuildTaskResource.State getTaskState(BuildIdentifier targetBuild, String taskPath) {
-        return buildControllerFor(targetBuild).getTaskState(taskPath);
```
❓Would it make sense to introduce a build controller for the root build instead perhaps?

jjohannes replied:
I have tried that. But the current interface and implementation also assume that the controller is responsible for "launching the build" (i.e. doing the task graph calculation) on a separate thread. While the main build is already in the state of doing that on the main thread.

I think we should clean this all up in a follow up. But it would be a bigger refactoring.
